source/docs/design/rw2022-talk-proposal.tex

```diff
@@ -0,0 +1,199 @@
+Chris Hawblitzel -- Microsoft Research \\
+Bryan Parno, Yi Zhou, Chanhee Cho, Travis Hance -- Carnagie Mellon University \\
+Jon Howell -- VMware Research}
+
+\begin{document}
+
+\maketitle
+
+
+This talk will introduce Verus, an SMT-based tool we are building to
+verify full functional correctness of systems code written in Rust. I am
+planning to demonstrate Verus via live demos interspersed in the talk:
+the examples in this proposal are samples of the snippets that I'll
+verify with Verus in the demos.
+
+\section{Introduction}\label{introduction}
+
+The Rust programming language, with its combination of high-level
+features, memory safety, and low-level ``unsafe'' primitives has proven
+to be an excellent tool to write high-performance systems software, with
+direct interactions with the hardware.
+
+We observed that functions using the safe subset of Rust without
+interior mutability are pure, deterministic transformations, even when
+they manipulate mutable references (as they can be logically rewritten
+as consuming the initial value of the reference, and returning the new
+value). By statically preventing aliased mutable references Rust
+relieves the burden on the developer of reasoning about data races; when
+verifying imperative code in languages that do not restric aliasing, the
+developer similarly has to explicitly manage the heap and potential
+aliasing. In our experience, a Rust-like linear type system can be an
+excellent aid to Floyd-Hoare SMT-based verification of executable code.
+
+We condensed these intuitions as the main design principle of our new
+tool for SMT-based verification of Rust code, Verus. We are building
+Verus to efficiently verify full functional correctness of low-level
+systems code written in a safe Rust dialect that supports expressing
+specifications and proofs.
+
+\section{Verus' design}\label{verus-design}
+
+Verus leverages the fact that safe Rust has a natural, direct encoding
+in SMT (for code without interior mutability, which we need to address
+separately). Algebraic data types (ADTs) and sound generics complement
+safe Rust to make it a solid basis for a mathematical language for
+expressing specifications and proofs. When functions represent theorems,
+expressing complex proofs like lemmas by induction is straightforward
+via a simple recursive call.
+
+Other tools, like Dafny, need complex encodings in SMT to support the
+semantics of the language (such as heap reasoning), and in our
+experience this can be the cause of unpredictable prover performance. We
+choose to prioritize efficient, straighforward encoding of Rust to Z3 in
+Verus (without an intermediate language like Boogie). This design
+principle comes with a necessary compromise: we do not intend to support
+all Rust features and libraries, instead excluding those that would
+require complicated encoding. We do however plan to provide
+verification-aided safe alternatives when unsafe is currently required
+in Rust.
```

Comment:
I felt like "when" was a little unclear

maybe something like "for patterns where unsafe would otherwise be required"?